File: Sources/Claudario/Server/EventRouter.swift
import Foundation

enum HookEvent: String {
    case sessionStart      = "SessionStart"
    case userPromptSubmit  = "UserPromptSubmit"
    case preToolUse        = "PreToolUse"
    case postToolUse       = "PostToolUse"
    case notification      = "Notification"
    case stop              = "Stop"
    case subagentStop      = "SubagentStop"
    case sessionEnd        = "SessionEnd"
}

final class EventRouter {
    private let onWalk: () -> Void
    private let onIdle: () -> Void
    private let onCelebrate: () -> Void
    private let onNotify: () -> Void

    private struct SessionState {
        var isWalking: Bool
        var lastSeen: Date
    }
    private var sessions: [String: SessionState] = [:]
    private let lock = NSLock()
    private let staleAfter: TimeInterval = 5 * 60

    init(onWalk: @escaping () -> Void,
         onIdle: @escaping () -> Void,
         onCelebrate: @escaping () -> Void,
         onNotify: @escaping () -> Void) {
        self.onWalk = onWalk
        self.onIdle = onIdle
        self.onCelebrate = onCelebrate
        self.onNotify = onNotify
    }

    func handle(_ json: Data) {
        guard let obj = (try? JSONSerialization.jsonObject(with: json)) as? [String: Any] else { return }
        let eventName = (obj["hook_event_name"] as? String) ?? ""
        let sessionId = (obj["session_id"] as? String) ?? "default"
        if let event = HookEvent(rawValue: eventName) {
            process(event: event, sessionId: sessionId)
        } else if eventName == "PermissionRequest" {
            DispatchQueue.main.async { self.onNotify() }
        }
    }

    private func process(event: HookEvent, sessionId: String) {
        lock.lock()
        evictStaleLocked()
        defer { lock.unlock() }

        let now = Date()
        switch event {
        case .sessionStart:
            sessions[sessionId] = SessionState(isWalking: false, lastSeen: now)

        case .userPromptSubmit, .preToolUse:
            let wasWalking = anyWalkingLocked
            sessions[sessionId] = SessionState(isWalking: true, lastSeen: now)
            if !wasWalking {
                DispatchQueue.main.async { self.onWalk() }
            }

        case .postToolUse:
            sessions[sessionId]?.lastSeen = now

        case .notification:
            sessions[sessionId]?.lastSeen = now
            DispatchQueue.main.async { self.onNotify() }

        case .stop, .subagentStop:
            sessions[sessionId]?.isWalking = false
            sessions[sessionId]?.lastSeen = now
            if !anyWalkingLocked {
                DispatchQueue.main.async {
                    self.onIdle()
                    self.onCelebrate()
                }
            }

        case .sessionEnd:
            sessions[sessionId] = nil
            if !anyWalkingLocked {
                DispatchQueue.main.async { self.onIdle() }
            }
        }
    }

    private var anyWalkingLocked: Bool {
        sessions.values.contains { $0.isWalking }
    }

    private func evictStaleLocked() {
        let cutoff = Date().addingTimeInterval(-staleAfter)
        sessions = sessions.filter { $0.value.lastSeen >= cutoff }
    }

    func reset() {
        lock.lock()
        sessions.removeAll()
        lock.unlock()
        DispatchQueue.main.async { self.onIdle() }
    }
}

File: windows/Claudario.Windows/Server/EventRouter.cs
using System.Text.Json;
using Claudario.Windows.Mascot;

namespace Claudario.Windows.Server;

public sealed class EventRouter
{
    public Action? OnWalk      { get; set; }
    public Action? OnIdle      { get; set; }
    public Action? OnCelebrate { get; set; }
    public Action? OnNotify    { get; set; }
    public Action<MascotActivity>? OnActivity { get; set; }

    private struct SessionState
    {
        public bool           IsWalking;
        public MascotActivity Activity;
        public DateTime       LastSeen;
    }

    private readonly Dictionary<string, SessionState> _sessions = new();
    private readonly object _lock        = new();
    private string?         _lastActiveSessionId;
    private MascotActivity  _lastEmitted = MascotActivity.Idle;
    private static readonly TimeSpan StaleAfter = TimeSpan.FromMinutes(5);

    public void Handle(byte[] json)
    {
        JsonDocument doc;
        try { doc = JsonDocument.Parse(json); }
        catch { return; }

        using (doc)
        {
            var root = doc.RootElement;
            string eventName = root.TryGetProperty("hook_event_name", out var en) ? en.GetString() ?? "" : "";
            string sessionId = root.TryGetProperty("session_id",       out var si) ? si.GetString() ?? "default" : "default";
            string? toolName = root.TryGetProperty("tool_name",        out var tn) ? tn.GetString() : null;

            Process(eventName, sessionId, toolName);
        }
    }

    private void Process(string eventName, string sessionId, string? toolName)
    {
        lock (_lock)
        {
            EvictStale();

            var now  = DateTime.UtcNow;
            var sess = _sessions.TryGetValue(sessionId, out var s) ? s
                     : new SessionState { IsWalking = false, Activity = MascotActivity.Idle, LastSeen = now };
            sess.LastSeen = now;
            _lastActiveSessionId = sessionId;

            switch (eventName)
            {
                case "SessionStart":
                    sess.IsWalking = false;
                    sess.Activity  = MascotActivity.Idle;
                    break;

                case "UserPromptSubmit":
                {
                    bool wasWalking = AnyWalking();
                    sess.IsWalking = true;
                    sess.Activity  = MascotActivity.Thinking;
                    _sessions[sessionId] = sess;
                    if (!wasWalking) FireOnMain(OnWalk);
                    EmitActivity();
                    return;
                }

                case "PreToolUse":
                {
                    bool wasWalking = AnyWalking();
                    sess.IsWalking = true;
                    sess.Activity  = MascotActivityExtensions.CategoryForTool(toolName ?? "");
                    _sessions[sessionId] = sess;
                    if (!wasWalking) FireOnMain(OnWalk);
                    EmitActivity();
                    return;
                }

                case "PostToolUse":
                    sess.Activity = MascotActivity.Thinking;
                    _sessions[sessionId] = sess;
                    EmitActivity();
                    return;

                case "PreCompact":
                    sess.IsWalking = true;
                    sess.Activity  = MascotActivity.Compacting;
                    _sessions[sessionId] = sess;
                    FireOnMain(OnWalk);
                    EmitActivity();
                    return;

                case "Notification" or "PermissionRequest":
                    FireOnMain(OnNotify);
                    break;

                case "Stop" or "SubagentStop":
                    sess.IsWalking = false;
                    sess.Activity  = MascotActivity.Idle;
                    _sessions[sessionId] = sess;
                    if (!AnyWalking())
                    {
                        FireOnMain(OnIdle);
                        FireOnMain(OnCelebrate);
                        EmitIdleActivity();
                    }
                    else
                    {
                        EmitActivity();
                    }
                    return;

                case "SessionEnd":
                    _sessions.Remove(sessionId);
                    if (!AnyWalking())
                    {
                        FireOnMain(OnIdle);
                        EmitIdleActivity();
                    }
                    else
                    {
                        EmitActivity();
                    }
                    return;
            }

            _sessions[sessionId] = sess;
        }
    }

    private void EmitActivity()
    {
        MascotActivity activity;
        if (_lastActiveSessionId is not null &&
            _sessions.TryGetValue(_lastActiveSessionId, out var ls) && ls.IsWalking)
            activity = ls.Activity;
        else if (_sessions.Values.FirstOrDefault(x => x.IsWalking) is { IsWalking: true } w)
            activity = w.Activity;
        else
            activity = MascotActivity.Idle;

        if (activity == _lastEmitted) return;
        _lastEmitted = activity;
        var act = activity;
        FireOnMain(() => OnActivity?.Invoke(act));
    }

    private void EmitIdleActivity()
    {
        if (_lastEmitted == MascotActivity.Idle) return;
        _lastEmitted = MascotActivity.Idle;
        FireOnMain(() => OnActivity?.Invoke(MascotActivity.Idle));
    }

    private bool AnyWalking() => _sessions.Values.Any(s => s.IsWalking);

    private void EvictStale()
    {
        var cutoff = DateTime.UtcNow - StaleAfter;
        foreach (var key in _sessions.Keys.ToList())
            if (_sessions[key].LastSeen < cutoff)
                _sessions.Remove(key);
    }

    public void Reset()
    {
        bool needsIdleEmit;
        lock (_lock)
        {
            _sessions.Clear();
            _lastActiveSessionId = null;
            needsIdleEmit = _lastEmitted != MascotActivity.Idle;
            _lastEmitted  = MascotActivity.Idle;
        }
        FireOnMain(OnIdle);
        if (needsIdleEmit) FireOnMain(() => OnActivity?.Invoke(MascotActivity.Idle));
    }

    private static void FireOnMain(Action? action)
    {
        if (action is null) return;
        System.Windows.Application.Current?.Dispatcher.BeginInvoke(action);
    }
}
